src/libraries/System.Security.Cryptography/tests/X509Certificates/X509FilesystemTests.Windows.cs

```diff
@@ -0,0 +1,464 @@
+        {
+            EnsureNoKeysGained(
+                (Flags: storageFlags, Bytes: MakePfx(1, 5, capi)),
+                state => X509CertificateLoader.LoadPkcs12(state.Bytes, "", state.Flags, s_cspPreservingLimits));
+        }
+
+        [Theory]
+        [InlineData(X509KeyStorageFlags.DefaultKeySet)]
+        [InlineData(X509KeyStorageFlags.DefaultKeySet, true)]
+        [InlineData(X509KeyStorageFlags.UserKeySet)]
+        [InlineData(X509KeyStorageFlags.UserKeySet, true)]
+        [InlineData(X509KeyStorageFlags.MachineKeySet)]
+        [InlineData(X509KeyStorageFlags.MachineKeySet, true)]
+        public static void AllFilesDeleted_MultiplePrivateKey_CollLoader_KeepCsp(X509KeyStorageFlags storageFlags, bool capi = false)
+        {
+            EnsureNoKeysGained(
+                (Flags: storageFlags, Bytes: MakePfx(6, capi: capi)),
+                state => new ImportedCollection(
+                    X509CertificateLoader.LoadPkcs12Collection(state.Bytes, "", state.Flags, s_cspPreservingLimits)));
+        }
+
+        [Theory]
+        [InlineData(X509KeyStorageFlags.DefaultKeySet)]
+        [InlineData(X509KeyStorageFlags.DefaultKeySet, true)]
+        [InlineData(X509KeyStorageFlags.UserKeySet)]
+        [InlineData(X509KeyStorageFlags.UserKeySet, true)]
+        [InlineData(X509KeyStorageFlags.MachineKeySet)]
+        [InlineData(X509KeyStorageFlags.MachineKeySet, true)]
+        public static void AllFilesDeleted_SinglePrivateKey_CollLoader_KeepCsp(X509KeyStorageFlags storageFlags, bool capi = false)
+        {
+            EnsureNoKeysGained(
+                (Flags: storageFlags, Bytes: MakePfx(1, 5, capi)),
+                state => new ImportedCollection(
+                    X509CertificateLoader.LoadPkcs12Collection(state.Bytes, "", state.Flags, s_cspPreservingLimits)));
+        }
+
+        private static byte[] MakePfx(
+            int certAndKeyCount = 1,
+            int certOnlyCount = 0,
+            bool capi = false,
+            [CallerMemberName] string? name = null)
+        {
+            Pkcs12SafeContents keys = new Pkcs12SafeContents();
+            Pkcs12SafeContents certs = new Pkcs12SafeContents();
+            DateTimeOffset notBefore = DateTimeOffset.UtcNow.AddMinutes(-5);
+            DateTimeOffset notAfter = notBefore.AddMinutes(10);
+
+            // Every other key is a machine key. Depending on the minute when this line of code runs, it's
+            // either starting with the first, or with the second.
+            int machineKeySkew = DateTime.UtcNow.Minute % 2;
+
+            PbeParameters pbeParams = new PbeParameters(
+                PbeEncryptionAlgorithm.TripleDes3KeyPkcs12,
+                HashAlgorithmName.SHA1,
+                1);
+
+            for (int i = 0; i < certAndKeyCount; i++)
+            {
+                using (RSA key = RSA.Create(1024))
```

Comment:
Well, it's only running on Windows, so other platforms being upset shouldn't be an issue.  I was going to use ECDSA since their keygen is faster, but then CAPI can't be tested; so it was really just about keeping keygen relatively fast.  But I can time it to see if 2048 adds noticeable delay.

Reply:
It was pretty noticeable (10s -> 30s for just this class, and similar increment for the full library suite).  So I made a set of 6 random keys for the class instead of 6 per test-variant.  Now keygen is negligible compared to the rest of the work, so bumped the size.